devito/passes/iet/mpi.py

```diff
@@ -48,157 +49,155 @@ def _drop_halospots(iet):
+                    hs1.halo_scheme.fmapper[f] = hse
+
+                    hsmapper[hs1] = hsmapper.get(hs1, hs1.halo_scheme).drop(f)
+                    imapper[it].append(hs1.halo_scheme.project(f))
 
-    # Post-process analysis
     mapper = {i: HaloSpot(i._rebuild(), HaloScheme.union(hss))
               for i, hss in imapper.items()}
+
     mapper.update({i: i.body if hs.is_void else i._rebuild(halo_scheme=hs)
                    for i, hs in hsmapper.items()})
-
     # Transform the IET hoisting/dropping HaloSpots as according to the analysis
     iet = Transformer(mapper, nested=True).visit(iet)
-
-    # Clean up: de-nest HaloSpots if necessary
-    mapper = {}
-    for hs in FindNodes(HaloSpot).visit(iet):
-        if hs.body.is_HaloSpot:
-            halo_scheme = HaloScheme.union([hs.halo_scheme, hs.body.halo_scheme])
-            mapper[hs] = hs._rebuild(halo_scheme=halo_scheme, body=hs.body.body)
-    iet = Transformer(mapper, nested=True).visit(iet)
-
     return iet
 
 
 def _merge_halospots(iet):
     """
     Merge HaloSpots on the same Iteration tree level where all data dependencies
-    would be honored.
-    """
-
-    # Merge rules -- if the retval is True, then it means the input `dep` is not
-    # a stopper to halo merging
-
-    def rule0(dep, hs, loc_indices):
-        # E.g., `dep=W<f,[t1, x]> -> R<f,[t0, x-1]>` => True
-        return not any(d in hs.dimensions or dep.distance_mapper[d] is S.Infinity
-                       for d in dep.cause)
+    would be honored. Helps to avoid redundant halo exchanges when the same data is
+    redundantly exchanged within the same Iteration tree level as well as to initiate
+    multiple halo exchanges at once.
 
-    def rule1(dep, hs, loc_indices):
-        # TODO This is apparently never hit, but feeling uncomfortable to remove it
-        return (dep.is_regular and
-                dep.read is not None and
-                all(not any(dep.read.touched_halo(d.root)) for d in dep.cause))
+    Example:
 
-    def rule2(dep, hs, loc_indices):
-        # E.g., `dep=W<f,[t1, x+1]> -> R<f,[t1, xl+1]>` and `loc_indices={t: t0}` => True
-        return any(dep.distance_mapper[d] == 0 and dep.source[d] is not v
-                   for d, v in loc_indices.items())
+    for time                       for time
+      haloupd v[t0]                  haloupd v[t0], h
+      W v[t1]- R v[t0]               W v[t1]- R v[t0]
+      haloupd v[t0], h
+      W g[t1]- R v[t0], h            W g[t1]- R v[t0], h
 
-    rules = [rule0, rule1, rule2]
+    """
 
     # Analysis
-    cond_mapper = MapHaloSpots().visit(iet)
-    cond_mapper = {hs: {i for i in v if i.is_Conditional and
-                        not isinstance(i.condition, GuardFactorEq)}
-                   for hs, v in cond_mapper.items()}
+    cond_mapper = _make_cond_mapper(iet)
+
+    mapper = {}
 
     iter_mapper = MapNodes(Iteration, HaloSpot, 'immediate').visit(iet)
```

Comment:
now that I think about it, you can probably also move the `cond_mapper` inside such a new utility function, and do the conditional filtering in there directly, rather than repeating it in both passes (it's basically the same stuff repeated at the moment)

Reply:
I think this can be resolved